### ISSUE TITLE
Batch normalization in ReLU

### ISSUE THREAD

**Issue by redst4r** [Lasagne/Lasagne#635]
Hi,

lasagne.layers.batch_norm conveniently removes the bias term from the layer, as it is redundant with the batchnorms mean subtraction.

However, I recently read [here](http://arxiv.org/abs/1507.02672) (page 6), that in a similar way, for ReLUs the scaling in the Batchnorm is redundant:

> For example, the rectified linear unit does not need scaling, the linear activation function needs neither scaling nor bias, but softmax requires both.

First, I'm not totally sure if its true ( they have a funny nomenclature, and according to their formulas, they use scaling/bias slightly differently). I tried to do the math myself, but couldn't definitely figure out if its correct. Intuitively, it makes sense: You shift your input to zero mean; everything <0 gets squashed to 0 by the RelU; all >0 gets just the identity, and its irrelevant how you scale this.

Can we handle/remove this redundany somehow in batch_norm? Since the ReLU is the default nonlinearity for most (?) layers, this might change quite a bit

**Comment by benanne** [comment]
Don't forget that you still need some form of scaling after the last ReLU layer though. You can disable the scaling factor by passing `gamma=None` when you instantiate a batch norm layer (then it will be fixed to 1).

**Comment by f0k** [comment]
> Don't forget that you still need some form of scaling after the last ReLU layer though.

Usually the last layer is not ReLU, so for most use cases it should be fine to have the `batch_norm()` macro pass `gamma=None` if the nonlinearity is relu (either non-leaky relu or prelu) and the user didn't pass `gamma`. This would change the number of parameters for existing models trained with batch normalization, though, so it would partly break backwards compatibility. I don't know if it's worth this...

**Comment by benanne** [comment]
I don't think adding that kind of special-case logic is a good idea. Lasagne is geared towards researchers and I think we should assume that researchers know what they are doing. If they are not, then that is their problem, not ours ;)

**Comment by f0k** [comment]
Okay, let's leave this up to the users then. @redst4r, note that you can just write your own `batch_norm()` macro that calls Lasagne's `batch_norm()` macro with `gamma=None` if the given layer has a rectifier nonlinearity, so it's just as easy to use as now. Closing!
